ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/worldstate/MarkSweepPruner.java

```diff
@@ -0,0 +1,185 @@
+import tech.pegasys.pantheon.services.kvstore.KeyValueStorage.Transaction;
+import tech.pegasys.pantheon.util.bytes.Bytes32;
+import tech.pegasys.pantheon.util.bytes.BytesValue;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class MarkSweepPruner {
+  private static final Logger LOG = LogManager.getLogger();
+  private static final BytesValue IN_USE = BytesValue.of(1);
+  private static final int MARKS_PER_TRANSACTION = 1000;
+  private final WorldStateStorage worldStateStorage;
+  private final KeyValueStorage markStorage;
+  private final Counter markedNodesCounter;
+  private final Counter markOperationCounter;
+  private final Counter sweepOperationCounter;
+  private final Counter sweptNodesCounter;
+  private long nodeAddedListenerId;
+  private final ReentrantLock markLock = new ReentrantLock(true);
+  private final Set<BytesValue> pendingMarks = Collections.newSetFromMap(new ConcurrentHashMap<>());
+
+  public MarkSweepPruner(
+      final WorldStateStorage worldStateStorage,
+      final KeyValueStorage markStorage,
+      final MetricsSystem metricsSystem) {
+    this.worldStateStorage = worldStateStorage;
+    this.markStorage = markStorage;
+
+    markedNodesCounter =
+        metricsSystem.createCounter(
+            PantheonMetricCategory.PRUNER,
+            "marked_nodes_total",
+            "Total number of nodes marked as in use");
+    markOperationCounter =
+        metricsSystem.createCounter(
+            PantheonMetricCategory.PRUNER,
+            "mark_operations_total",
+            "Total number of mark operations performed");
+
+    sweptNodesCounter =
+        metricsSystem.createCounter(
+            PantheonMetricCategory.PRUNER,
+            "swept_nodes_total",
+            "Total number of unused nodes removed");
+    sweepOperationCounter =
+        metricsSystem.createCounter(
+            PantheonMetricCategory.PRUNER,
+            "sweep_operations_total",
+            "Total number of sweep operations performed");
+  }
+
+  public void prepare() {
+    worldStateStorage.removeNodeAddedListener(nodeAddedListenerId); // Just in case.
+    nodeAddedListenerId = worldStateStorage.addNodeAddedListener(this::markNewNodes);
+  }
+
+  // Note chainHeadStateRoot must be the state root of the current chain head.
+  // We can delay the actual sweep until a certain number of blocks in the future if we want to
+  // have a certain number of block history available.
+  // TODO: Need to ensure we only start marking when new world states aren't being persisted
+  // Once we add our node added listener persisting world states can continue but we can't start
+  // half way through persisting a world state.
```

Comment:
@ajsutton Is some of this outdated? As far as my understanding goes, the hash given to mark needn't be the chain head, is that right?

Any new thoughts on the persistence?

Reply:
The block we mark needs to be the current chain head at the time that we add the NodeAddedListener.  We can then wait a few blocks before we actually start marking to ensure we aren't on a short lived fork - it's adding the NodeAddedListener while the mark block is head that's critical.

Otherwise if head was 10 and we picked 5 as the mark block and added the NodeAddedListener, blocks 11 onwards would be marked by the NodeAddedListener and 5 would be marked by walking its trie but 6,7,8,9 would be missed.